### PR TITLE
MM-53687: Block changes to name, display name or purpose for direct and group messages

### LINKED ISSUE #24162: Updating direct or group messages info breaks channel list
[https://github.com/mattermost/mattermost/issues/24023](https://github.com/mattermost/mattermost/issues/24023|smart-link) 

#### Summary

When [API request](https://github.com/mattermost/mattermost/issues/%7B%7BbaseUrl%7D%7D/api/v4/channels/:channel_id) updates channel info for private conversation, you are able to update the channel name. Doing so, makes the channel show as empty row and not selectable

#### Steps to reproduce

Create private conversation
Update any info by endpoint `PUT {{baseUrl`/api/v4/channels/:channel_id}} with the following payload:

```json
{
    "id": {{channel_id}},
    "name": "test",
    "display_name": "test",
    "purpose": "test",
    "header": "test"
}
```

Conversation is broken

#### Expected behavior

1. Server block changes to name, display name or purpose for Direct and Group messages
1. Server says '400 Bad Request' and add more details to the response
1. The channel does not get updated

#### Observed behavior

Video: [https://user-images.githubusercontent.com/81349914/253495561-37880c20-e539-4d9a-9956-235c80660a66.mp4](https://user-images.githubusercontent.com/81349914/253495561-37880c20-e539-4d9a-9956-235c80660a66.mp4)

#### Possible fixes

Changes on the API layer  (`server/channels/api4/channel.go`) around the rest of the checks done in this file to follow the behaviour described in the Expected behaviour section.

----
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-53687

### PR DESCRIPTION
**Pull request by kyeongsoosoo**
#### Summary

Block changes to name, display name or purpose for direct and group messages

#### Ticket Link

Fixes #24162 
https://mattermost.atlassian.net/browse/MM-53687

#### Release Note

```release-note
Block changes to name, display name or purpose for direct and group messages
```

### CODE REVIEW

**Comment by larkox** [comment]
@kyeongsoosoo Do you mind adding tests for this?

**Comment by kyeongsoosoo** [comment]
> @kyeongsoosoo Do you mind adding tests for this?

I've added the test!

**Review by larkox — APPROVED** [review]

File: server/channels/api4/channel.go [hunk 1]
```diff
@@ -169,6 +169,10 @@ func updateChannel(c *Context, w http.ResponseWriter, r *http.Request) {
 			c.Err = model.NewAppError("updateChannel", "api.channel.patch_update_channel.forbidden.app_error", nil, "", http.StatusForbidden)
 			return
 		}
+		if channel.Name != "" || channel.DisplayName != "" || channel.Purpose != "" {
+			c.Err = model.NewAppError("updateChannel", "api.channel.update_channel.cant_update_direct_or_group_messages.app_error", nil, "", http.StatusBadRequest)
+			return
+		}
```
Comment (agnivade) [inline server/channels/api4/channel.go:175]: This is wrong and assumes that the user should only set those fields which are meant to be changed. That's what patchChannel is for. In `UpdateChannel`, users can set the old fields as unchanged as well.

We need to compare the new channel value with the old channel one, like in L187 and then throw an error accordingly. And we also need to update the `patchChannel` endpoint.

**Review by agnivade — CHANGES_REQUESTED** [review]

File: server/channels/api4/channel.go (hunk 1 above)
Comment (larkox) [reply server/channels/api4/channel.go:175]: That seems to be my misunderstanding laying out the description of the ticket. @kyeongsoosoo do you mind updating the PR according to these comments? Thanks!

Comment (kyeongsoosoo) [reply server/channels/api4/channel.go]: Thanks for the review. 
I've updated the PR based on your comments.

File: server/channels/api4/channel.go [hunk 2]
```diff
@@ -169,6 +169,10 @@ func updateChannel(c *Context, w http.ResponseWriter, r *http.Request) {
 			c.Err = model.NewAppError("updateChannel", "api.channel.patch_update_channel.forbidden.app_error", nil, "", http.StatusForbidden)
 			return
 		}
+		if (channel.Name != "" && channel.Name != oldChannel.Name) || (channel.DisplayName != "" && channel.DisplayName != oldChannel.DisplayName) || (channel.Purpose != oldChannel.Purpose) {
```
Comment (agnivade) [inline server/channels/api4/channel.go:172]: Why can't we change the purpose? @larkox

File: server/i18n/en.json [hunk 3]
```diff
@@ -335,6 +335,10 @@
     "id": "api.channel.patch_channel_moderations.license.error",
     "translation": "Your license does not support channel moderation"
   },
+  {
+    "id": "api.channel.patch_update_channel.cant_update_direct_or_group_messages.app_error",
+    "translation": "Unable to update the name, display_name, and purpose of a direct or group messages."
```
Comment (agnivade) [inline server/i18n/en.json:340]: "Unable" is not appropriate here. It means something went wrong in trying to update the field. It should rather be "You are not allowed to update the ..."

File: server/channels/api4/channel.go (hunk 2 above)
Comment (larkox) [reply server/channels/api4/channel.go:172]: That would be a question for product. :P

As far as I know, there is no way on the UI to change or see the purpose of a DM/GM, so... it shouldn't be allowed to change.

Comment (agnivade) [reply server/channels/api4/channel.go:172]: I think it should be fine. The server in general should be more accepting of custom clients doing their own thing. If there is no reason for a limitation, it should be fine to allow clients to change fields that the UI doesn't support. 

@johndavidlugtu - do we want to intentionally block clients from updating the `purpose` field for DM/GMs?

File: server/i18n/en.json (hunk 3 above)
Comment (kyeongsoosoo) [reply server/i18n/en.json]: I updated the message. I'll update the message if the `purpose` field becomes updatable.

File: server/channels/api4/channel.go (hunk 2 above)
Comment (johndavidlugtu) [reply server/channels/api4/channel.go:172]: @agnivade I'm leaning towards keeping things consistent and not allowing ability to update the purpose field of a DM/GM.

Rationale being, as we move towards making clearer the distinction between a DM/GM vs channels, we want to reduce overlap where possible. DMs and GMs are supposed to be ephemeral and/or used less frequently compared to channels.

As we currently don't allow in the UI to change the purpose of DM/GM, let's keep it as is. That said, if we hear feedback from users, we can reconsider in the future.

Still open to thoughts if you have anything more to add

Comment (agnivade) [reply server/channels/api4/channel.go:172]: Got it. Sounds good to me.

File: server/i18n/en.json [hunk 4]
```diff
@@ -335,6 +335,10 @@
     "id": "api.channel.patch_channel_moderations.license.error",
     "translation": "Your license does not support channel moderation"
   },
+  {
+    "id": "api.channel.patch_update_channel.update_direct_or_group_messages_not_allowed.app_error",
```
Comment (agnivade) [inline server/i18n/en.json:339]: `purpose of direct or group messages`

File: server/i18n/en.json [hunk 5]
```diff
@@ -419,6 +423,10 @@
     "id": "api.channel.restore_channel.unarchived",
     "translation": "{{.Username}} unarchived the channel."
   },
+  {
+    "id": "api.channel.update_channel.update_direct_or_group_messages_not_allowed.app_error",
+    "translation": "You are not allowed to update the name, display_name, and purpose of a direct or group messages."
```
Comment (agnivade) [inline server/i18n/en.json:428]: Same as above

File: server/i18n/en.json (hunk 4 above)
Comment (kyeongsoosoo) [reply server/i18n/en.json:339]: I fixed it. Thank you

**Review by johndavidlugtu — APPROVED** [review]

**Review by agnivade — APPROVED** [review]

**Comment by agnivade** [comment]
@kyeongsoosoo - There is a linting failure due to shadowing. Should be good to go once you fix that.

**Comment by kyeongsoosoo** [comment]
> @kyeongsoosoo - There is a linting failure due to shadowing. Should be good to go once you fix that.

@agnivade  I fIxed it.

**Comment by larkox** [comment]
@yasserfaraazkhan QA notes:

This is for when we try to update the channel purpose / header through the API directly. You should try to call directly the update channel API with different values to check that you cannot change the name/display name/purpose of DMs/GMs but you can still change the name/display name/purpose of private and open channels.

**Review by yasserfaraazkhan — APPROVED** [review]

**Pull request merged**